src/Microsoft.VisualStudio.Composition/ReflectionHelpers.cs

```diff
@@ -677,6 +691,40 @@ internal static int GetMetadataTokenSafe(this MemberInfo memberInfo)
             return memberInfo.MetadataToken;
         }
 
+        internal static bool TryGetContainerType(Type type, out TypeInfo containerType)
+        {
+            Requires.NotNull(type, nameof(type));
+
+            // Discover the ICollection<T> or ICollection<Lazy<T, TMetadata>> interface implemented by this type.
+            var icollectionTypes =
+                from iface in new[] { type }.Concat(type.GetTypeInfo().ImplementedInterfaces)
+                let ifaceInfo = iface.GetTypeInfo()
+                where ifaceInfo.IsGenericType
+                let genericTypeDef = ifaceInfo.GetGenericTypeDefinition()
+                where genericTypeDef.Equals(typeof(ICollection<>)) || genericTypeDef.Equals(typeof(IEnumerable<>)) || genericTypeDef.Equals(typeof(IList<>))
+                select ifaceInfo;
+
+            containerType = icollectionTypes.FirstOrDefault();
+
+            return containerType != null;
+        }
+
+        internal static Type GetElementType(Type type)
```

Comment:
This appears to be functionally equivalent to `PartDiscovery.GetElementTypeFromMany`. Can we share code instead of reimplementing it?